scvi/external/cellassign/_module.py

```diff
@@ -0,0 +1,225 @@
+        Number of categories for each extra categorical covariate
+    n_continuous_cov
+        Number of continuous covarites
+    """
+
+    def __init__(
+        self,
+        n_genes: int,
+        rho: torch.Tensor,
+        n_batch: int = 0,
+        n_cats_per_cov: Optional[Iterable[int]] = None,
+        n_continuous_cov: int = 0,
+    ):
+        super().__init__()
+        self.n_genes = n_genes
+        self.n_labels = rho.shape[1]
+        self.n_batch = n_batch
+        self.n_cats_per_cov = n_cats_per_cov
+        self.n_continuous_cov = n_continuous_cov
+
+        # this would be P in their code
+        # if it's zero, let's ignore
+        # though we do need the base expression intercept either way
+        design_matrix_col_dim = n_batch + n_continuous_cov
+        design_matrix_col_dim += 0 if n_cats_per_cov is None else sum(n_cats_per_cov)
+        self.register_buffer("rho", rho)
+
+        # perform all other initialization
+        self.min_delta = 2
+        dirichlet_concentration = torch.tensor([1e-2] * self.n_labels)
+        self.register_buffer("dirichlet_concentration", dirichlet_concentration)
+        self.shrinkage = True
+        self.b_g_0 = torch.nn.Parameter(torch.randn(n_genes))
+
+        # compute theta
+        self.theta_logit = torch.nn.Parameter(torch.randn(self.n_labels))
+
+        # compute delta (cell type specific overexpression parameter)
+        self.delta_log = torch.nn.Parameter(
+            torch.FloatTensor(self.n_genes, self.n_labels).uniform_(
+                np.log(self.min_delta), 2
+            )
+        )
+
+        # shrinkage prior on delta
+        if self.shrinkage:
+            self.delta_log_mean = torch.nn.Parameter(torch.Tensor(0))
+            self.delta_log_variance = torch.nn.Parameter(torch.Tensor(1))
+
+        self.log_a = torch.nn.Parameter(torch.zeros(B, dtype=torch.float64))
+
+    def _get_inference_input(self, tensors):
+        return {}
+
+    def _get_generative_input(self, tensors, inference_outputs):
+        x = tensors[_CONSTANTS.X_KEY]
+
+        to_cat = []
+        if self.n_batch > 0:
+            to_cat.append(one_hot(tensors[_CONSTANTS.BATCH_KEY], self.n_batch))
+
+        cont_key = _CONSTANTS.CONT_COVS_KEY
+        if cont_key in tensors.keys():
+            to_cat.append(tensors[cont_key])
+
+        cat_key = _CONSTANTS.CAT_COVS_KEY
+        if cat_key in tensors.keys():
+            for cat_input, n_cat in zip(
+                torch.split(tensors[cat_key], 1, dim=1), self.n_cats_per_cov
+            ):
+                to_cat.append(one_hot(cat_input, n_cat))
+
+        design_matrix = torch.cat(to_cat) if len(to_cat) > 0 else None
+
+        input_dict = dict(x=x, design_matrix=design_matrix)
+        return input_dict
+
+    @auto_move_data
+    def inference(self):
+        return {}
+
+    @auto_move_data
+    def generative(self, x, design_matrix=None):
+        # x has shape (n, g)
+        delta = torch.exp(self.delta_log)  # (g, c)
```

Comment:
before this line, you should create a clamped delta, https://pytorch.org/docs/stable/generated/torch.clamp.html

clamp delta_log with a minimum value of log (min_delta)